brushfire-core/src/main/scala/com/stripe/brushfire/TreeTraversal.scala

```diff
@@ -93,78 +179,59 @@ object TreeTraversal {
-          order.reverse
-      }
-
-    loop(0D, SortedMap.empty, Vector.empty, xs).toList
-  }
-
-  def mkRandom(id: String): Random =
-    new Random(MurmurHash3.stringHash(id))
+  def probabilisticWeightedDepthFirst[Tree, K, V, T, A](implicit treeOps: FullBinaryTreeOps[Tree, BranchLabel[K, V, A], LeafLabel[T, A]], conversion: A => Double): TreeTraversal[Tree, K, V, T, A] =
+    DepthFirstTreeTraversal(Reorder.probabilisticWeightedDepthFirst(conversion))
 }
 
-case class DepthFirstTreeTraversal[K, V, T, A](order: (Random, List[Node[K, V, T, A]]) => List[Node[K, V, T, A]])
-    extends TreeTraversal[K, V, T, A] {
-
-  def find(start: Node[K, V, T, A], row: Map[K, V], id: Option[String]): Stream[LeafNode[K, V, T, A]] = {
-    // Lazy to avoid creation in the fast case.
-    lazy val rng: Random = id.fold[Random](Random)(TreeTraversal.mkRandom)
-
-    // A little indirection makes scalac happy to eliminate some tailcalls in loop.
-    def loop0(stack: List[Node[K, V, T, A]]): Stream[LeafNode[K, V, T, A]] =
-      loop(stack)
-
-    @tailrec
-    def loop(stack: List[Node[K, V, T, A]]): Stream[LeafNode[K, V, T, A]] =
-      stack match {
-        case Nil =>
-          Stream.empty
-        case (leaf @ LeafNode(_, _, _)) :: rest =>
-          leaf #:: loop0(rest)
-        case (split @ SplitNode(_)) :: rest =>
-          val newStack = split.findChildren(row) match {
-            case Nil => rest
-            case node :: Nil => node :: rest
-            case candidates => order(rng, candidates) ::: rest
-          }
-          loop(newStack)
+case class DepthFirstTreeTraversal[Tree, K, V, T, A](reorder: Reorder[A])(implicit val treeOps: FullBinaryTreeOps[Tree, BranchLabel[K, V, A], LeafLabel[T, A]]) extends TreeTraversal[Tree, K, V, T, A] {
+
+  import treeOps.{Node, foldNode}
+
+  def searchNode(start: Node, row: Map[K, V], id: Option[String]): Stream[LeafLabel[T, A]] = {
+
+    // this will be a noop unless we have an id and our reorder
+    // instance requires randomness. it ensures that each searchNode
+    // call has its own independent RNG (in cases where we care about
+    // repeatability, i.e. when `id` is not None).
+    val r = reorder.setSeed(id)
+
+    // pull the A value out of a branch or leaf.
+    val getAnnotation: Node => A =
+      n => foldNode(n)((_, _, bl) => bl._3, ll => ll._3)
+
+    // construct a singleton stream from a leaf
+    //val Empty: Stream[LeafLabel[T, A]] = Stream.empty
+    val leafF: LeafLabel[T, A] => Stream[LeafLabel[T, A]] =
+      _ #:: Stream.empty
+
+    // recurse into branch nodes, going left, right, or both,
+    // depending on what our predicate says.
+    lazy val branchF: (Node, Node, BranchLabel[K, V, A]) => Stream[LeafLabel[T, A]] =
+      { case (lc, rc, (k, p, _)) =>
+        p(row.get(k)) match {
+          case Some(true) =>
+            recurse(lc)
+          case Some(false) =>
+            recurse(rc)
+          case None =>
+            val (c1, c2) = r(lc, rc, getAnnotation)
```

Comment:
To be clear, re-order wouldn't be handling the traversal, just the ordering, but instead of returning a `tuple` from `apply`, we'd take 1 more function argument:

```
def apply[B](n1: N, n2: N, a: A)(f: (N, N) => B): B
```

Is that what you were thinking too?